### PR TITLE
Optimize extract_dll_paths in ninja backend

### PR DESCRIPTION
**Pull request by bruchar1**
On my project, this function was a bottleneck in setup time.

By properly caching it, the cumtime for this function went from 26s to 4.3s when I profiled it.

### CODE REVIEW

File: mesonbuild/backend/backends.py [hunk 1]
```diff
@@ -1116,32 +1144,9 @@ def extract_dll_paths(cls, target: build.BuildTarget) -> T.Set[str]:
...
-
-                else:
-                    if p.is_file():
-                        p = p.parent
-                    # Heuristic: replace *last* occurence of '/lib'
-                    binpath = Path('/bin'.join(p.as_posix().rsplit('/lib', maxsplit=1)))
-                    for _ in binpath.glob('*.dll'):
-                        results.add(str(binpath))
-                        break
+            results.update(filter(None, map(cls.search_dll_path, dep.link_args)))  # pylint: disable=bad-builtin
```
Comment (tristan957) [inline mesonbuild/backend/backends.py:1149]: Should this just be a list comprehension instead?

This looks like 2 iterations of a `T.Iterable` that could be done in one.

**Comment by tristan957** [comment]
Nice find @bruchar1!

File: mesonbuild/backend/backends.py (hunk 1 above)
Comment (bruchar1) [reply mesonbuild/backend/backends.py:1149]: Huh? Do you mean : `results.update(dll_path for link_args in dep.link_args for dll_path in cls.search_dll_path(link_args) if dll_path is not None)` ? It seems more obscure to me...

Comment (bruchar1) [reply mesonbuild/backend/backends.py:1149]: And since those are all generators, it is really just one chained iteration....

Comment (tristan957) [reply mesonbuild/backend/backends.py:1149]:
```python
for arg in dep.link_args:
  dll_path = cls.search_dll_path(arg)
  if dll_path is not None:
    results.append(dll_path)
```

Probably no good way for list comprehension.

Comment (bruchar1) [reply mesonbuild/backend/backends.py:1149]: Personally, I still prefer `filter` and `map`... This is probably a matter of taste.

**Comment by bruchar1** [comment]
Could this one go into 1.2.1? I think a performance fix can be considered as a bugfix...

**Comment by tristan957** [comment]
I added it to the 1.3.0 milestone. Someone else can decide on on the 1.2.1 thing

**Pull request merged**